[SPARK-14251][SQL] Add SQL command for printing out generated code for debugging

dongjoon-hyun opened the pull request:
## What changes were proposed in this pull request?

This PR implements `EXPLAIN CODEGEN` SQL command which returns generated codes like `debugCodegen`. In `spark-shell`, we don't need to `import debug` module. In `spark-sql`, we can use this SQL command now.

**Before**

```
scala> import org.apache.spark.sql.execution.debug._
scala> sql("select 'a' as a group by 1").debugCodegen()
Found 2 WholeStageCodegen subtrees.
== Subtree 1 / 2 ==
...

Generated code:
...

== Subtree 2 / 2 ==
...

Generated code:
...
```

**After**

```
scala> sql("explain extended codegen select 'a' as a group by 1").collect().foreach(println)
[Found 2 WholeStageCodegen subtrees.]
[== Subtree 1 / 2 ==]
...
[]
[Generated code:]
...
[]
[== Subtree 2 / 2 ==]
...
[]
[Generated code:]
...
```
## How was this patch tested?

Pass the Jenkins tests (including new testcases)

dongjoon-hyun committed:
[SPARK-14251][SQL] Add SQL command for printing out generated code for debugging

SparkQA commented:
**[Test build #54669 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54669/consoleFull)** for PR 12099 at commit [`4e750e5`](https://github.com/apache/spark/commit/4e750e5cfecc8b1ac70c788c5811bc947d8acccb).

hvanhovell commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -576,7 +576,7 @@ frameBound
 
 explainOption
-    : LOGICAL | FORMATTED | EXTENDED
+    : LOGICAL | FORMATTED | EXTENDED | CODEGEN
```
`CODEGEN` is not a keyword. Please add to the `nonReserved` rule.

rxin replied:
@hvanhovell does this fix the problem already? i wasn't sure what was before your comment vs after.

hvanhovell replied:
Yeah it is fixed. He added it to the nonReserved rule (see below).

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -122,9 +122,10 @@ class SparkSqlAstBuilder extends AstBuilder {
+        codegen = options.exists(_.CODEGEN != null))
     } else {
-      ExplainCommand(OneRowRelation)
+      ExplainCommand(OneRowRelation, codegen = options.exists(_.CODEGEN != null))
```
is there a point showing the generated code of an otherwise unexplainable command?

hvanhovell commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveExplainSuite.scala:
```diff
@@ -101,4 +101,19 @@ class HiveExplainSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
     }
   }
+
+  test("SPARK-14251: EXPLAIN CODEGEN command") {
```
Is it feasible to add an actual codegenerated example? Or do you think this is too long/prone to changes?

rxin replied:
it's really long -- i think having the first couple lines are good enough.

hvanhovell replied:
Ok.

Regarding the name of the test. I would typically only add the JIRA number to test name if it fixes a bug (that needs some explanation).

rxin replied:
yea i agree. it's best not to have the jira ticket here

dongjoon-hyun replied:
Thanks. Now, it's changed to **EXPLAIN CODEGEN command**.

hvanhovell commented:
@dongjoon-hyun I left a few small comments. Looks pretty solid overall.

dongjoon-hyun commented:
Thank you, @hvanhovell .

dongjoon-hyun committed:
According to the comments, update the code.

SparkQA commented:
**[Test build #54673 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54673/consoleFull)** for PR 12099 at commit [`83c3c01`](https://github.com/apache/spark/commit/83c3c01f8c938fb5247045d98e187294dabfcf8f).

dongjoon-hyun commented:
Hi, @hvanhovell .
I updated the PR according to your comments. For the real generated code, I think it's safe to maintain the first three lines since that will not change frequently.

SparkQA commented:
**[Test build #54669 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54669/consoleFull)** for PR 12099 at commit [`4e750e5`](https://github.com/apache/spark/commit/4e750e5cfecc8b1ac70c788c5811bc947d8acccb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54669/
Test PASSed.

SparkQA commented:
**[Test build #54673 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54673/consoleFull)** for PR 12099 at commit [`83c3c01`](https://github.com/apache/spark/commit/83c3c01f8c938fb5247045d98e187294dabfcf8f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54673/
Test PASSed.

dongjoon-hyun commented:
Hi, @rxin .
Could you review this PR?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:
```diff
@@ -81,29 +100,11 @@ package object debug {
-      }
-      output
-    }
+    def debugCodegenString(): String = codegenString(query.queryExecution.executedPlan)
```
do we still need this function?

dongjoon-hyun replied:
It is still used in `DebuggingSuite` since I didn't change that testcase.
```
val res = sqlContext.range(10).groupBy("id").count().debugCodegenString()
```

rxin replied:
yea let's change that

dongjoon-hyun replied:
Sure! No problem at all :)

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala:
```diff
@@ -237,15 +237,18 @@ case class ExplainCommand(
       Seq(AttributeReference("plan", StringType, nullable = true)()),
-    extended: Boolean = false)
+    extended: Boolean = false,
+    codegen: Boolean = false)
```
maybe we should have a separate command? otherwise we are going to add a lot of parameters.

dongjoon-hyun replied:
Oh, do you mean `codegen select ...` instead `explain codegen select ...`?
I thought `codegen` was an boolean option flag, but new command is also possible if it need more other options.
What kind of usage for `codegen` command do you think? Could you enumerate some for me?

rxin replied:
i like the sql -- just saying it'd be great to have different command classes rather than having different flags in a single command.

dongjoon-hyun replied:
As you know, I just implemented this because SPARK-14251 describes `explain codegen select * ...`.
I thought SPARK-14251 aims to reduce the debugging efforts by showing the full info in a single command.
If that is not a goal, we should change it. 

You are the issue reporter of SPARK-14251. :)
If you want to do this in another way, I can change this right now. This is the very right time. :)

Could you update [SPARK-14251](https://issues.apache.org/jira/browse/SPARK-14251) for me before making `CODEGEN` command?

Thank you, @rxin .

rxin replied:
explain codegen should just dump the same thing debugCodegen dumps. Nothing else.

dongjoon-hyun replied:
Sure. I'll change that. By the way, I think you mean (1). Am I right?
(1) `EXPLAIN CODEGEN SELECT ~`
(2) `CODEGEN SELECT ~`

rxin replied:
yes explain codegen

dongjoon-hyun replied:
Thank you. I do it right now.

dongjoon-hyun committed:
Remove `debugCodegenString` function

dongjoon-hyun commented:
Hi, @rxin . `debugCodegenString` is removed and `DebuggingSuite` is modified accordingly.

SparkQA commented:
**[Test build #54713 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54713/consoleFull)** for PR 12099 at commit [`11efbbb`](https://github.com/apache/spark/commit/11efbbb587bfb844248e445e688b7f6819126c68).

dongjoon-hyun committed:
`EXPLAIN CODEGEN` should be equal to `debugCodegen`.

SparkQA commented:
**[Test build #54717 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54717/consoleFull)** for PR 12099 at commit [`497a93e`](https://github.com/apache/spark/commit/497a93ec34eb22655a9dabce5c08879fc9e1d30f).

dongjoon-hyun commented:
Now, `EXPLAIN CODEGEN` is the logically same with `debugCodegen`. I updated the description of this PR, too.

dongjoon-hyun commented:
One minor thing, why @DeepSparkBot does not visit this PR? 
Today, @DeepSparkBot  came to my other PR. :)

SparkQA commented:
**[Test build #54713 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54713/consoleFull)** for PR 12099 at commit [`11efbbb`](https://github.com/apache/spark/commit/11efbbb587bfb844248e445e688b7f6819126c68).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54713/
Test PASSed.

andrewor14 commented:
You need to say: deep-review this please

dongjoon-hyun commented:
Oh, I see. Thank you, @andrewor14 !

dongjoon-hyun committed:
Update according to the DeepSparkBot.

SparkQA commented:
**[Test build #54717 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54717/consoleFull)** for PR 12099 at commit [`497a93e`](https://github.com/apache/spark/commit/497a93ec34eb22655a9dabce5c08879fc9e1d30f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54717/
Test PASSed.

SparkQA commented:
**[Test build #54736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54736/consoleFull)** for PR 12099 at commit [`7b10a51`](https://github.com/apache/spark/commit/7b10a517e04c72235beb9cfa271dc540df75b763).

dongjoon-hyun commented:
I updated the code according to @DeepSparkBot 's comments, and also rebased to the master since master branch is now Scala 2.11.8.
It took sometime to test locally.

SparkQA commented:
**[Test build #54736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54736/consoleFull)** for PR 12099 at commit [`7b10a51`](https://github.com/apache/spark/commit/7b10a517e04c72235beb9cfa271dc540df75b763).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54736/
Test FAILed.

SparkQA commented:
**[Test build #2728 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2728/consoleFull)** for PR 12099 at commit [`7b10a51`](https://github.com/apache/spark/commit/7b10a517e04c72235beb9cfa271dc540df75b763).

dongjoon-hyun committed:
Remove JIRA-ID in testcase name.

SparkQA commented:
**[Test build #54759 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54759/consoleFull)** for PR 12099 at commit [`9334fa9`](https://github.com/apache/spark/commit/9334fa9930222912cd847911eb04eb90da9253de).

SparkQA commented:
**[Test build #2728 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2728/consoleFull)** for PR 12099 at commit [`7b10a51`](https://github.com/apache/spark/commit/7b10a517e04c72235beb9cfa271dc540df75b763).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented:
Thanks - going to merge this in master.

dongjoon-hyun commented:
Thank you, @rxin , @hvanhovell , @andrewor14 , and @DeepSparkBot . :)

SparkQA commented:
**[Test build #54759 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54759/consoleFull)** for PR 12099 at commit [`9334fa9`](https://github.com/apache/spark/commit/9334fa9930222912cd847911eb04eb90da9253de).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54759/
Test PASSed.
